src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs

```diff
@@ -155,37 +156,47 @@ public bool ValidateProcessedBlock(Block processedBlock, TxReceipt[] receipts, B
 
     private bool ValidateWithdrawals(Block block, IReleaseSpec spec, out string? error)
     {
-        if (spec.WithdrawalsEnabled && block.Withdrawals is null)
+        static bool Fail(ILogger logger, string error)
         {
-            error = $"Withdrawals cannot be null in block {block.Hash} when EIP-4895 activated.";
-
-            if (_logger.IsWarn) _logger.Warn(error);
-
+            if (logger.IsWarn) logger.Warn(error);
             return false;
         }
 
-        if (!spec.WithdrawalsEnabled && block.Withdrawals is not null)
+        if (spec.WithdrawalsEnabled && block.Withdrawals is null)
         {
-            error = $"Withdrawals must be null in block {block.Hash} when EIP-4895 not activated.";
-
-            if (_logger.IsWarn) _logger.Warn(error);
+            return Fail(_logger, error = $"Withdrawals cannot be null in block {block.ToString(Block.Format.FullHashAndNumber)} when EIP-4895 activated.");
+        }
 
-            return false;
+        if (!spec.WithdrawalsEnabled && block.Withdrawals is not null)
+        {
+            return Fail(_logger, error = $"Withdrawals must be null in block {block.ToString(Block.Format.FullHashAndNumber)} when EIP-4895 not activated.");
         }
 
         if (block.Withdrawals is not null)
         {
             if (!ValidateWithdrawalsHashMatches(block, out Keccak withdrawalsRoot))
             {
-                error = $"Withdrawals root hash mismatch in block {block.ToString(Block.Format.FullHashAndNumber)}: expected {block.Header.WithdrawalsRoot}, got {withdrawalsRoot}";
-                if (_logger.IsWarn) _logger.Warn($"Withdrawals root hash mismatch in block {block.ToString(Block.Format.FullHashAndNumber)}: expected {block.Header.WithdrawalsRoot}, got {withdrawalsRoot}");
+                return Fail(_logger, error = $"Withdrawals root hash mismatch in block {block.ToString(Block.Format.FullHashAndNumber)}: expected {block.Header.WithdrawalsRoot}, got {withdrawalsRoot}");
+            }
 
-                return false;
+            // We should initialize it from parent?
+            ulong expectedIndex = block.Withdrawals.FirstOrDefault()?.Index ?? 0;
```

Comment:
we're starting from 0 according to EIP
"a monotonically increasing index, starting from 0," Not sure if FirstOrDefault is needed

Reply:
Yes, but starting from 0 is between blocks. In a block it starts at some index (that is index in last block +1)

Reply:
yes, agreed. I observed it on devnet